XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,419 @@
+the same as initially requested, but the user will be able to achieve the same
+end result.
+
+Performance impact of `AsyncWrap` should be zero if not being used, and near
+zero while being used. The performance overhead of `AsyncHook` callbacks
+supplied by the user should account for essentially all the performance
+overhead.
+
+
+## Terminology
+
+Because of the potential ambiguity for those not familiar with the terms "handle"
+and "request" they should be well defined.
+
+Handles are a reference to a system resource. The resource handle can be a
+simple identifier, such as a file descriptor, or it can be a pointer that
+allows access to further information, such as `uv_tcp_t`. Realize that the
+latter still relies on a file descriptor, but has been encapsulated in a data
+structure containing additional information that allows the application to
+work with it.
+
+Requests are short lived data structures created to accomplish one task. The
+callback for a request should always and only ever fire one time. When the
+assigned task has either completed or encountered an error. Requests are used
+by handles to perform work. Such as accepting a new connection or writing data
+to disk.
+
+
+## API
+
+### Overview
+
+```js
+// Standard way of requiring a module. Snake case follows core module
+// convention.
+const async_wrap = require('async_wrap');
+
+// List of asynchronous subsystems (e.g. CRYPTO or TCPWRAP). Can be used for
+// trace filtering of events.
+const subsystems = async_wrap.subsystems;
+
+// Retrieve a handle by its id. The id will have been passed to the init()
+// callback. Reason for this API is to prevent users from holding references
+// to the handles themselves. Thus possibly preventing GC from cleaning up
+// the handle and ever firing the destroy() callback.
+const handle = async_wrap.getHandleById(n);
+
+// init() is called during object construction. The handle will have not
+// completed construction when this callback runs. So not all fields will have
+// populated.
+function init(id, type, parentId) { }
+
+// pre() is called just before the handle's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWRAP), and should be called exactly 1
+// time for requests (e.g. FSREQWRAP).
+function pre(id) { }
+
+// post() is called just after the handle's callback has finished, and will be
+// called regardless whether the handle's callback threw. If the handle's
+// callback did throw then hasThrown will be true.
+function post(id, hasThrown) { }
+
+// destroy() is called when an AsyncWrap instance is destroyed. In cases like
+// HTTPPARSER where the resource is reused, or timers where the handle is only
+// a JS object, destroy() will be triggered manually soon after post() has
+// completed.
+function destroy(id) { }
+
+// Create a new instance and add each callback to the corresponding hook.
+var asyncHook = new AsyncHook({ init, pre, post, destroy });
```

Comment:
Forgive me if this is naïve, as `AsyncWrap` has been in my peripheral vision mostly, but is `AsyncHook` a new global? Or is it `require('async_wrap`).AsyncHook`? If the latter, should that be explicitly indicated?